// @strapi/design-system/dist/Card/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const Card = require('./Card.js');
const CardAction = require('./CardAction.js');
const CardAsset = require('./CardAsset.js');
const CardBadge = require('./CardBadge.js');
const CardBody = require('./CardBody.js');
const CardCheckbox = require('./CardCheckbox.js');
const CardContent = require('./CardContent.js');
const CardHeader = require('./CardHeader.js');
const CardTimer = require('./CardTimer.js');
const CardTitle = require('./CardTitle.js');



exports.Card = Card.Card;
exports.CardAction = CardAction.CardAction;
exports.CardAsset = CardAsset.CardAsset;
exports.CardBadge = CardBadge.CardBadge;
exports.CardBody = CardBody.CardBody;
exports.CardCheckbox = CardCheckbox.CardCheckbox;
exports.CardContent = CardContent.CardContent;
exports.CardHeader = CardHeader.CardHeader;
exports.CardTimer = CardTimer.CardTimer;
exports.CardSubtitle = CardTitle.CardSubtitle;
exports.CardTitle = CardTitle.CardTitle;
